Update redux for protocols & sessions

bryfox opened the pull request:
Resolves #545. Resolves #558.

This updates redux actions (with some minor changes to state shape) to better support protocol exporting and simplify some of the import/load/set flows with sessions. This builds on the FinishSession interface branch (a merge should wait on #559), but the two can be reviewed in parallel.

This also adds support for loading devtools extensions in electron (as redux-devtools was very useful for this effort), but installation left as a manual exercise (see Readme).

### Notes on changes
- Removes some of the differences between `loadFactoryProtocol` and `loadProtocol`
  + In particular, both now make use of `addSession()`
  + ...and session IDs are no longer generated by protocol actions
  + This reduces some of the coupling, but interfaces must call addSession in both cases now. (Unwanted side-effects from programming error should be solved by #570.)
- Removes the `remoteId` that was passed through actions (to provide initial support for export). Remote IDs are now calculated locally as they are not server-specific.

### Manual testing

We don't have much integration testing yet, and there's still integration between sessions and protocols. I've gone through combinations of {protocolType} + {loadAction|exportAction}:

- Factory protocols vs Downloaded (Server- and URL-sourced) protocols
- Loading
  - existing sessions
    - via app reload
    - via 'resume interview'
  - new Sessions
    - via protocol cards (startup screen)
    - via download flows
- Exporting (or not) as expected

rebeccamadsen commented:
If I try to "import from url" in chrome (which shouldn't work, but I can click on it, so), a "fake" session gets created.

Some protocol names disappear on the "resume interview" screen. I think when I add a protocol over again from the server list, or from the "import from url" button.

Otherwise the iterations I tried worked well. I like the list for manual testing you've started.

wwqrd commented on public/electron-starter.js:
```diff
@@ -125,12 +125,21 @@ function createMenu() {
   Menu.setApplicationMenu(appMenu);
 }
 
+function loadDevToolsExtensions() {
```
Cool! I'll add this to Architect.

bryfox commented:
> If I try to "import from url" in chrome (which shouldn't work, but I can click on it, so), a "fake" session gets created.

Good catch! This is related to #569, where fake sessions are created if 'load' fails. But in this branch, they'll also be created if 'download' fails. I think the same problem would occur if, for example, you entered an import URL that 404'd.

This was a side-effect of my using `addSession()` for download protocols (not just factory). I'll need to do a little more thinking on a fix.

bryfox commented:
I've rebased this onto master, which now contains the FinishSession interface.

I addressed the two issues @rebeccamadsen found as follows:

1. 'Fake' Sessions with no corresponding protocol are removed from the SessionList. For now, I'm letting them be created in the store; they're a rare case. Removing them entirely will take more integration between protocols & sessions, which should be cleaner once we get rid of the 'factory' special case. We can use #569 to track.

2. Sessions whose protocolPath ref is outdated (indicating the protocol was updated) display '[version out of date]' instead of the name. Displaying the name for these would again require more coupling between the modules, and (in the near future) protocol updates will invalidate/remove old sessions anyway.
